substrate/frame/staking/stake-tracker/src/lib.rs

```diff
@@ -0,0 +1,830 @@
+								.defensive_proof(
+									"staker exists in the list as per the check above; qed.",
+								);
+							false
+						}
+					} else {
+						defensive!("unexpected: unable to fetch score from staking interface of an existent staker");
+						false
+					}
+				},
+			};
+
+			log!(
+				debug,
+				"update_score of {:?} by {:?}. removed target node? {}",
+				who,
+				imbalance,
+				removed
+			);
+		}
+	}
+}
+
+#[cfg(any(test, feature = "try-runtime"))]
+impl<T: Config> Pallet<T> {
+	/// Try-state checks for the stake-tracker pallet.
+	///
+	/// 1. `do_try_state_approvals`: checks the curent approval stake in the target list compared
+	///    with the staking state.
+	/// 2. `do_try_state_target_sorting`: checks if the target list is sorted by score.
+	pub fn do_try_state() -> Result<(), sp_runtime::TryRuntimeError> {
+		#[cfg(feature = "try-runtime")]
+		Self::do_try_state_target_sorting()?;
+		Self::do_try_state_approvals()
+	}
+
+	/// Try-state: checks if the approvals stake of the targets in the target list are correct.
+	///
+	/// These try-state checks generate a map with approval stake of all the targets based on
+	/// the staking state of stakers in the voter and target lists. In doing so, we are able to
+	/// verify that the current voter and target lists and scores are in sync with the staking
+	/// data and perform other sanity checks as the approvals map is calculated.
+	///
+	/// NOTE: this is an expensive state check since it iterates over all the nodes in the
+	/// target and voter list providers.
+	///
+	/// Invariants:
+	///
+	/// * Target List:
+	///   * The sum of the calculated approvals stake is the same as the current approvals in
+	///   the target list per target.
+	///   * The target score of an active validator is the sum of all of its nominators' stake
+	///   and the self-stake;
+	///   * The target score of an idle validator (i.e. chilled) is the sum of its nominator's
+	///   stake. An idle target may not be part of the target list, if it has no nominations.
+	///   * The target score of a "dangling" target (ie. idle AND unbonded validator) must
+	///   always be > 0. We expect the stake-tracker to have cleaned up dangling targets with 0
+	///   score.
+	///   * The number of target nodes in the target list matches the number of
+	///   (active_validators + idle_validators + dangling_targets_score_with_score).
+	///
+	/// * Voter List:
+	///  * The voter score is the same as the active stake of the corresponding stash.
+	///  * An active validator should also be part of the voter list.
+	///  * An idle validator should not be part of the voter list.
+	///  * A dangling target shoud not be part of the voter list.
+	pub(crate) fn do_try_state_approvals() -> Result<(), sp_runtime::TryRuntimeError> {
+		let mut approvals_map: BTreeMap<AccountIdOf<T>, ExtendedBalance> = BTreeMap::new();
+
+		// build map of approvals stakes from the `VoterList` POV.
+		for voter in T::VoterList::iter() {
```

Comment:
In this case, we want to rebuild the target list from based on the voters in the voter list. This will also allow us to do some sanity checks along the way, which is helpful.

Reply:
By `VoterList / TargetList` I meant the difference of those sets. If someone is in the target list, that means they are a validator and they could not be nominators, right? Which would mean that any account which is in the voter list but isn't in the target list is a nominator. Is this wrong @gpestana ?

Assuming that holds: everything else that you do for nominator sanity checks is guarded by `if let Some(nominations) = <T::Staking as StakingInterface>::nominations(&voter)`. My suggestion would be to do something like
```
for voter in T::VoterList::iter().collect::<BTreeSet<_>>.difference(T::TargetList::iter().collect::<BTreeSet<_>>()) {
    let nomination = <T::Staking as StakingInterface>::nominations(&voter).expect("must be a nominator")
    ...
}
```

So a more explicit check on the nominator status instead of a possible silent failure with `if let Some(...)`.

Reply:
as per our discussion offline, I added a couple more checks that cover all our cases https://github.com/paritytech/polkadot-sdk/pull/1933/commits/ff8ee33885fd9533f66d7af35bfdb9d4b16cd5ba

thanks!